File: csharp/18_error_handling.cs
// ============================================================
// Topic:  Error Handling
// Summary: try/catch, throw, custom Exception, finally, wrap/rethrow.
// Run:     dotnet script csharp/18_error_handling.cs
//
// KEY DIFFERENCES:
//   C#:     try/catch/finally, throw, Exception hierarchy, when filter
//   Python: try/except/finally, raise, custom Exception subclass
//   TS/JS:  try/catch/finally, throw, Error subclass
//   Rust:   Result<T,E>, ? operator, panic! — no exceptions by default
//   Java:   try/catch/finally, throw, checked vs unchecked exceptions
// ============================================================

using System;

// ----- 1. Safe path — no exception -----
int SafeDivide(int a, int b) => a / b;
Console.WriteLine($"result: {SafeDivide(20, 2)}");     // result: 10 | normal return

// ----- 2. try/catch — handle DivideByZeroException -----
try
{
    SafeDivide(10, 0);                               // throws    | int division by zero
}
catch (DivideByZeroException)
{
    Console.WriteLine("caught: division by zero");   // caught    | specific handler
}

// ----- 3. Custom exception — throw and catch by type -----
class CustomBadException : Exception
{
    public CustomBadException(string message) : base(message) { }
}

try
{
    throw new CustomBadException("invalid");         // throw     | Python: raise CustomBad("invalid")
}
catch (CustomBadException ex)
{
    Console.WriteLine($"CustomBad: {ex.Message}");   // CustomBad: invalid
}

// ----- 4. finally — always runs (success or failure) -----
try
{
    int n = SafeDivide(30, 3);                       // succeeds  | finally still executes
    _ = n;
}
catch (DivideByZeroException)
{
    // not reached
}
finally
{
    Console.WriteLine("finally ran");                // finally ran | cleanup / guaranteed path
}

// ----- 5. Wrap and rethrow — inner fails, outer catches wrapped message -----
void RiskyWork()
{
    throw new InvalidOperationException("detail");   // inner     | original failure
}

try
{
    try
    {
        RiskyWork();
    }
    catch (Exception ex)
    {
        throw new ApplicationException("inner failed", ex);  // wrap | preserve InnerException
    }
}
catch (ApplicationException ex)
{
    Console.WriteLine($"outer: {ex.Message}");       // outer: inner failed | outer handler
}

// Cross-language note:
//   catch (Exception ex) catches all managed exceptions; order matters — most specific first.
//   `throw;` rethrows preserving stack trace; `throw ex;` resets it — prefer bare throw.
//   finally runs even with return/break in try; use using/IDisposable for resource cleanup too.
//   ApplicationException.InnerException holds the original — unwrap for full diagnostic chain.

File: cpp/18_error_handling.cpp
// ============================================================
// Topic:  Error Handling  (FOCUS: try/catch, throw, std::exception, RAII finally)
// Summary: Exceptions with std::exception hierarchy — scope guard simulates finally.
// Run:     g++ -std=c++17 cpp/18_error_handling.cpp -o cpp/18_error_handling && cpp/18_error_handling
//
// KEY DIFFERENCES:
//   C++:     try/catch/throw, std::exception, RAII for cleanup (no finally keyword)
//   Python:  try/except/finally, raise, custom Exception subclasses
//   Java:    try/catch/finally, throw, checked vs unchecked exceptions
//   C#:      try/catch/finally, throw, Exception base class
//   Rust:    Result<T,E> / ? operator — no exceptions by default
//   TS/JS:   try/catch/finally, throw, Error objects
// ============================================================

#include <exception>
#include <iostream>
#include <stdexcept>
#include <string>
#include <utility>

// ----- Custom exception — inherits std::exception (what() contract) -----
class CustomBad : public std::exception {
    std::string message_;

public:
    explicit CustomBad(std::string message) : message_(std::move(message)) {}

    const char* what() const noexcept override { return message_.c_str(); }
};

// ----- RAII scope guard — destructor runs on scope exit ("finally") -----
class FinallyGuard {
public:
    FinallyGuard() = default;
    ~FinallyGuard() { std::cout << "finally ran" << std::endl; }

    FinallyGuard(const FinallyGuard&) = delete;
    FinallyGuard& operator=(const FinallyGuard&) = delete;
};

int safe_divide(int numerator, int denominator) {
    if (denominator == 0) {
        throw std::runtime_error("division by zero");
    }
    return numerator / denominator;
}

int main() {
    // ----- 1. Successful path — no exception thrown -----
    try {
        std::cout << "result: " << safe_divide(20, 2) << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "unexpected: " << e.what() << std::endl;
    }
    // Output: result: 10

    // ----- 2. catch std::exception — store message from what() -----
    try {
        safe_divide(10, 0);
    } catch (const std::exception& e) {
        std::cout << "caught: " << e.what() << std::endl;
    }
    // Output: caught: division by zero

    // ----- 3. Custom exception + finally guard + rethrow to outer catch -----
    try {
        try {
            throw CustomBad("invalid");
        } catch (const CustomBad& e) {
            std::cout << "CustomBad: " << e.what() << std::endl;
            {
                FinallyGuard guard;  // RAII — prints on scope exit
            }
            throw std::runtime_error("inner failed");
        }
    } catch (const std::exception& e) {
        std::cout << "outer: " << e.what() << std::endl;
    }
    // Output: CustomBad: invalid
    // Output: finally ran
    // Output: outer: inner failed

    return 0;
}